File: life_app_back/internal/service/finance_service.go
package service

import (
	"life_app_back/internal/model"
	"life_app_back/internal/repository"
	"time"
)

// FinanceService 财务服务
type FinanceService struct {
	repo repository.FinanceRepository // 依赖注入 FinanceRepository
}

// NewFinanceService 创建 FinanceService 实例
func NewFinanceService() *FinanceService {
	return &FinanceService{
		repo: repository.FinanceRepository{},
	}
}

// AddTransaction 添加一笔交易记录
func (s *FinanceService) AddTransaction(userID uint, req model.AddTransactionRequest) error {
	// 构造 Transaction 模型
	transaction := model.Transaction{
		UserID:          userID,
		Type:            req.Type,
		Amount:          req.Amount,
		IconID:          req.IconID,
		Date:            req.Date,
		Merchant:        req.Merchant,
		Notes:           req.Notes,
		RecorderID:      req.RecorderID,
		IsFamilyExpense: req.IsFamilyExpense,
		ImageURL:        req.ImageURL,
		FamilyId:        req.FamilyId,
		GoalId:          req.GoalId,
	}

	// 调用 Repository 保存交易记录
	return s.repo.AddTransaction(&transaction)
}

// GetRecentTransactions 获取近期交易记录
func (s *FinanceService) GetRecentTransactions(userID uint, transactionType string, limit, memberId int, familyId int) ([]model.Transaction, error) {

	// 调用 Repository 获取近期交易记录
	return s.repo.GetRecentTransactions(userID, transactionType, limit, memberId, familyId)
}

// GetTransactions 获取交易记录列表，支持分页和筛选
func (s *FinanceService) GetTransactions(userID uint, params model.TransactionQueryParams) (*model.TransactionResponse, error) {
	// 确保用户ID被设置
	params.UserID = userID

	// 设置默认分页参数
	if params.Page <= 0 {
		params.Page = 1
	}
	if params.PageSize <= 0 {
		params.PageSize = 20
	}

	// 获取统计数据
	totalRecords, totalIncome, totalExpense, err := s.repo.GetTransactionStats(
		params.MemberID,
		params.StartDate,
		params.EndDate,
		params.Type,
		params.CategoryID,
		userID,
		params.FamilyId,
	)
	if err != nil {
		return nil, err
	}
	// 构造响应
	response := &model.TransactionResponse{
		Summary: struct {
			TotalCount   int     `json:"total_count"`
			TotalIncome  float64 `json:"total_income"`
			TotalExpense float64 `json:"total_expense"`
			NetAmount    float64 `json:"net_amount"`
		}{
			TotalCount:   totalRecords,
			TotalIncome:  totalIncome,
			TotalExpense: totalExpense,
			NetAmount:    totalIncome - totalExpense,
		},
	}

	return response, nil
}

// GetTransactionGroups 获取按日期分组的交易记录
func (s *FinanceService) GetTransactionGroups(userID uint, startDate, endDate time.Time, limit, page int, types string, memberID uint, categoryID []int, familyId uint) ([]model.TransactionDateGroup, error) {
	// 使用当前时间作为默认结束日期
	if endDate.IsZero() {
		endDate = time.Now()
	}

	// 如果未指定开始日期，则默认为30天前
	if startDate.IsZero() {
		startDate = endDate.AddDate(0, 0, -30)
	}

	// 使用默认limit
	if limit <= 0 {
		limit = 10
	}

	// 调用Repository方法
	return s.repo.GetTransactionsByDate(userID, startDate, endDate, limit, page, types, memberID, categoryID, familyId)
}

// GetTransactionTrend 获取交易趋势数据
func (s *FinanceService) GetTransactionTrend(userID uint, startDate, endDate time.Time, interval string, memberID uint, CategoryID []int, familyId uint) ([]model.TrendDataPoint, error) {

	// 调用Repository方法
	return s.repo.GetTransactionTrend(userID, startDate, endDate, interval, memberID, CategoryID, familyId)
}

// GetMemberExpenseStats 获取成员支出统计
func (s *FinanceService) GetMemberExpenseStats(userID uint, startDate, endDate time.Time) ([]model.MemberExpenseStats, error) {
	// 使用当前时间作为默认结束日期
	if endDate.IsZero() {
		endDate = time.Now()
	}

	// 如果未指定开始日期，则默认为30天前
	if startDate.IsZero() {
		startDate = endDate.AddDate(0, 0, -30)
	}

	// 调用Repository方法
	return s.repo.GetMemberExpenseStats(userID, startDate, endDate)
}

// GetFinanceSummary 获取财务摘要
func (s *FinanceService) GetFinanceSummary(userID uint, startDate, endDate time.Time) (map[string]interface{}, error) {
	// 使用当前时间作为默认结束日期
	if endDate.IsZero() {
		endDate = time.Now()
	}

	// 如果未指定开始日期，则默认为30天前
	if startDate.IsZero() {
		startDate = endDate.AddDate(0, 0, -30)
	}

	// 获取统计数据
	totalCount, totalIncome, totalExpense, err := s.repo.GetTransactionStats(0, startDate, endDate, "", []int{}, userID, 0)
	if err != nil {
		return nil, err
	}

	// 获取成员支出统计
	memberStats, err := s.repo.GetMemberExpenseStats(userID, startDate, endDate)
	if err != nil {
		return nil, err
	}

	// 获取每周趋势
	trendData, err := s.repo.GetTransactionTrend(userID, startDate, endDate, "week", 0, []int{}, 0)
	if err != nil {
		return nil, err
	}

	// 构造响应
	summary := map[string]interface{}{
		"totalCount":   totalCount,
		"totalIncome":  totalIncome,
		"totalExpense": totalExpense,
		"balance":      totalIncome - totalExpense,
		"memberStats":  memberStats,
		"trendData":    trendData,
		"periodStart":  startDate.Format("2006-01-02"),
		"periodEnd":    endDate.Format("2006-01-02"),
	}

	return summary, nil
}

// GetExpenseAnalysis 获取按类别统计的支出分析数据
func (s *FinanceService) GetExpenseAnalysis(userID uint, startDate, endDate time.Time, memberID uint, types string, familyId int) ([]model.CategoryExpenseStats, error) {
	// 使用当前时间作为默认结束日期
	if endDate.IsZero() {
		endDate = time.Now()
	}

	// 如果未指定开始日期，则默认为当月第一天
	if startDate.IsZero() {
		now := time.Now()
		startDate = time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, now.Location())
	}

	// 处理成员ID，如果未指定则获取所有家庭成员
	memberIDs := make([]uint, 0)
	if memberID == 0 && familyId > 0 {
		familyMemberRepository := &repository.FamilyMemberRepository{}

		familyMembers, err := familyMemberRepository.GetFamilyMembers(uint(familyId))
		if err != nil {
			return nil, err
		}
		for _, tmp := range familyMembers {
			memberIDs = append(memberIDs, tmp.ID)
		}
	}
	if memberID > 0 {
		memberIDs = append(memberIDs, memberID)
	}

	// 调用Repository方法获取按类别统计的支出数据
	return s.repo.GetCategoryExpenseStats(userID, startDate, endDate, memberIDs, types)
}

// GetMemberIncome 获取成员在指定时间段内的收入总额
func (s *FinanceService) GetMemberIncome(userID int64, startTime, endTime time.Time) (float64, error) {
	var totalIncome float64

	// 查询指定用户在时间段内的所有收入
	query := `
		SELECT COALESCE(SUM(amount), 0) 
		FROM transactions 
		WHERE user_id = ? AND type = 'income' AND date BETWEEN ? AND ?
	`

	err := model.DB.Raw(query, userID, startTime, endTime).Scan(&totalIncome).Error
	if err != nil {
		return 0, err
	}

	return totalIncome, nil
}

// GetMemberExpense 获取成员在指定时间段内的支出总额
func (s *FinanceService) GetMemberExpense(userID int64, startTime, endTime time.Time) (float64, error) {
	var totalExpense float64

	// 查询指定用户在时间段内的所有支出
	query := `
		SELECT COALESCE(SUM(amount), 0) 
		FROM transactions 
		WHERE user_id = ? AND type = 'expense' AND date BETWEEN ? AND ?
	`

	err := model.DB.Raw(query, userID, startTime, endTime).Scan(&totalExpense).Error
	if err != nil {
		return 0, err
	}

	return totalExpense, nil
}

// GetMembersFinanceData 批量获取多个成员的财务数据
// 返回一个map，key为成员ID，value为包含收入和支出的map
func (s *FinanceService) GetMembersFinanceData(userIDs []int64, startTime, endTime time.Time) (map[int64]map[string]float64, error) {
	// 调用Repository层的方法获取数据
	return s.repo.GetMembersFinanceData(userIDs, startTime, endTime)
}

File: life_app/lib/services/finance_service.dart
import 'package:flutter/material.dart';
import './api_service.dart'; // 导入 ApiService
import 'dart:convert'; // 用于 jsonEncode
import 'package:intl/intl.dart'; // 用于日期格式化

class FinanceService {
  final ApiService _apiService;

  // 恢复 BuildContext 参数，ApiService 需要它来获取认证令牌
  FinanceService() : _apiService = ApiService(); 

  /// 添加一笔交易记录
  Future<ApiResponse<void>> addTransaction({
    required BuildContext context, // 添加 context 参数用于获取认证令牌
    required String type, // "expense" or "income"
    required double amount,
    required int iconId,
    required DateTime date,
    String? merchant,
    String? notes,
    required int recorderId,
    bool? isFamilyExpense,
    String? imageUrl,
  }) async {
    try {
      final Map<String, dynamic> data = {
        'type': type,
        'amount': amount,
        'icon_id': iconId,
        'date': DateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", 'en_US').format(date.toUtc()), 
        'recorder_id': recorderId,
      };
      
      if (merchant != null && merchant.isNotEmpty) {
        data['merchant'] = merchant;
      }
      if (notes != null && notes.isNotEmpty) {
        data['notes'] = notes;
      }
      if (isFamilyExpense != null) {
        data['is_family_expense'] = isFamilyExpense;
      }
      if (imageUrl != null && imageUrl.isNotEmpty) {
        data['image_url'] = imageUrl;
      }

      print("Sending transaction data: ${jsonEncode(data)}");

      // 传递 context 以便 ApiService 可以获取认证令牌
      final responseData = await _apiService.post(
        path: '/api/v1/finance/transaction', 
        data: data,
        context: context, // 传递 context 参数
      );

      print("Received response data: $responseData"); 
      
      bool success = responseData['code'] == 0 || responseData['code'] == 200;
      String message = responseData['message'] ?? (success ? '添加成功' : '添加失败');

      return ApiResponse<void>(
        success: success,
        message: message,
        error: success ? null : (responseData['error'] ?? message),
      );

    } catch (e) {
      print("Error in FinanceService.addTransaction: $e"); 
      return ApiResponse<void>(
        success: false, 
        message: '发生未知错误: ${e.toString()}',
        error: e.toString(),
      );
    }
  }

  // 获取近期交易
  Future<ApiResponse> getRecentTransactions({
    required BuildContext context,
    required String type, // 'expense' 或 'income'
  }) async {
    try {
      final responseData = await _apiService.get(
        path: '/api/v1/finance/recent-transactions',
        params: {'type': type},
        context: context,
      );

      bool success = responseData['code'] == 0 || responseData['code'] == 200;
      return ApiResponse(
        success: success,
        data: success ? responseData['data'] : null,
        message: responseData['message'] ?? (success ? '获取成功' : '获取失败'),
      );
    } catch (e) {
      return ApiResponse(
        success: false,
        message: '获取近期交易失败: $e',
      );
    }
  }
  
  /// 获取交易统计摘要
  Future<ApiResponse> getTransactionSummary({
    required BuildContext context,
    DateTime? startDate,
    DateTime? endDate,
    String? type,
    int? memberId,
    int? categoryId,
    List<int>? categoryIds,
    int? familyId,
  }) async {
    try {
      print('开始获取交易摘要：');
      if (startDate != null) print('  开始日期：${DateFormat('yyyy-MM-dd HH:mm:ss').format(startDate)}');
      if (endDate != null) print('  结束日期：${DateFormat('yyyy-MM-dd HH:mm:ss').format(endDate)}');
      if (type != null) print('  交易类型：$type');
      if (memberId != null) print('  成员ID：$memberId');
      if (categoryId != null) print('  分类ID：$categoryId');
      if (categoryIds != null) print('  多选分类：$categoryIds');
      if (familyId != null) print('  家庭ID：$familyId');
      
      // 构建查询参数
      Map<String, String> params = {};
      
      // 添加日期范围参数
      if (startDate != null) {
        params['start_date'] = DateFormat('yyyy-MM-dd HH:mm:ss').format(startDate);
      }
      if (endDate != null) {
        params['end_date'] = DateFormat('yyyy-MM-dd HH:mm:ss').format(endDate);
      }
      
      // 添加其他可选参数
      if (type != null && type.isNotEmpty) {
        params['type'] = type;
      }
      if (memberId != null) {
        params['member'] = memberId.toString();
      }
      
      // 添加家庭ID参数
      if (familyId != null) {
        params['family_id'] = familyId.toString();
      }
      
      // 统一处理分类ID参数
      List<int> allCategoryIds = [];
      
      
      // 如果有多个分类ID，添加到列表
      if (categoryIds != null && categoryIds.isNotEmpty) {
        allCategoryIds.addAll(categoryIds.where((id) => id > 0));
      }
      
      // 处理分类ID参数
      if (allCategoryIds.isNotEmpty) {
        // 使用categories[]格式传递所有分类ID
        for (int i = 0; i < allCategoryIds.length; i++) {
          params['categories[$i]'] = allCategoryIds[i].toString();
        }
        print('  处理后的分类ID：$allCategoryIds');
      } else {
        print('  未找到有效的分类ID');
      }
      
      print('发送请求参数：$params');
      
      // 调用API
      final responseData = await _apiService.get(
        path: '/api/v1/finance/transactions',
        params: params,
        context: context,
      );
      
      bool success = responseData['code'] == 0 || responseData['code'] == 200;
      
      return ApiResponse(
        success: success,
        data: success ? responseData['data'] : null,
        message: responseData['message'] ?? (success ? '获取成功' : '获取失败'),
      );
    } catch (e) {
      print("Error in FinanceService.getTransactionSummary: $e");
      return ApiResponse(
        success: false,
        message: '获取交易统计摘要失败: $e',
      );
    }
  }
  
  /// 获取交易趋势数据
  Future<ApiResponse> getTransactionTrend({
    required BuildContext context,
    DateTime? startDate,
    DateTime? endDate,
    String interval = 'day', // 'day', 'week', 'month'
    String? type, // 可选参数，用于筛选特定类型：'expense' 或 'income'
    int? memberId, // 成员ID
    List<int>? categoryIds, // 分类ID列表
    int? familyId, // 家庭ID
  }) async {
    try {
      if (endDate == null) {
        endDate = DateTime.now();
      }
      
      // 根据interval设置默认的开始日期
      if (startDate == null) {
        switch (interval) {
          case 'day':
            startDate = endDate.subtract(const Duration(days: 30)); // 近30天
            break;
          case 'week':
            startDate = endDate.subtract(const Duration(days: 84)); // 近12周
            break;
          case 'month':
            startDate = DateTime(endDate.year - 1, endDate.month, 1); // 近12个月
            break;
          default:
            interval = 'day'; // 默认使用天为单位
            startDate = endDate.subtract(const Duration(days: 30)); // 近30天
        }
      }
      
      // 构建查询参数
      Map<String, String> params = {
        'start_date': DateFormat('yyyy-MM-dd HH:mm:ss').format(startDate),
        'end_date': DateFormat('yyyy-MM-dd HH:mm:ss').format(endDate),
        'interval': interval,
      };
      
      // 添加类型参数（如果有）
      if (type != null && type.isNotEmpty) {
        params['type'] = type;
      }
      
      // 添加成员ID（如果有）
      if (memberId != null) {
        params['member'] = memberId.toString();
      }
      
      // 添加家庭ID参数
      if (familyId != null) {
        params['family_id'] = familyId.toString();
      }
      
      // 处理分类ID参数
      if (categoryIds != null && categoryIds.isNotEmpty) {
        // 过滤掉无效的分类ID
        final validCategoryIds = categoryIds.where((id) => id > 0).toList();
        
        if (validCategoryIds.isNotEmpty) {
          // 使用categories[]格式传递所有分类ID
          for (int i = 0; i < validCategoryIds.length; i++) {
            params['categories[$i]'] = validCategoryIds[i].toString();
          }
          print('  分类ID：$validCategoryIds');
        }
      }
      
      print('发送请求参数：$params');
      
      // 调用API
      final responseData = await _apiService.get(
        path: '/api/v1/finance/trend',
        params: params,
        context: context,
      );
      
      bool success = responseData['code'] == 0 || responseData['code'] == 200;
      
      return ApiResponse(
        success: success,
        data: success ? responseData['data'] : null,
        message: responseData['message'] ?? (success ? '获取成功' : '获取失败'),
      );
    } catch (e) {
      print("Error in FinanceService.getTransactionTrend: $e");
      return ApiResponse(
        success: false,
        message: '获取交易趋势数据失败: $e',
      );
    }
  }

  /// 获取按日期分组的交易记录
  Future<ApiResponse> getTransactionGroups({
    required BuildContext context,
    DateTime? startDate,
    DateTime? endDate,
    String? type,
    int? memberId,
    List<int>? categoryIds,
    int? familyId,
    int limit = 20,
    int page = 1,
  }) async {
    try {
      print('开始获取日期分组的交易记录：');
      if (startDate != null) print('  开始日期：${DateFormat('yyyy-MM-dd HH:mm:ss').format(startDate)}');
      if (endDate != null) print('  结束日期：${DateFormat('yyyy-MM-dd HH:mm:ss').format(endDate)}');
      if (type != null) print('  交易类型：$type');
      if (memberId != null) print('  成员ID：$memberId');
      if (categoryIds != null) print('  分类IDs：$categoryIds');
      if (familyId != null) print('  家庭ID：$familyId');
      print('  页码：$page，每页数量：$limit');
      
      // 构建查询参数
      Map<String, String> params = {
        'limit': limit.toString(),
        'page': page.toString(),
      };
      
      // 添加日期范围参数
      if (startDate != null) {
        params['start_date'] = DateFormat('yyyy-MM-dd HH:mm:ss').format(startDate);
      }
      if (endDate != null) {
        params['end_date'] = DateFormat('yyyy-MM-dd HH:mm:ss').format(endDate);
      }
      
      // 添加其他可选参数
      if (type != null && type.isNotEmpty) {
        params['type'] = type;
      }
      if (memberId != null) {
        params['member'] = memberId.toString();
      }
      
      // 添加家庭ID参数
      if (familyId != null) {
        params['family_id'] = familyId.toString();
      }
      
      // 处理分类ID参数
      if (categoryIds != null && categoryIds.isNotEmpty) {
        // 过滤掉无效的分类ID
        final validCategoryIds = categoryIds.where((id) => id > 0).toList();
        
        if (validCategoryIds.isNotEmpty) {
          // 使用categories[]格式传递所有分类ID
          for (int i = 0; i < validCategoryIds.length; i++) {
            params['categories[$i]'] = validCategoryIds[i].toString();
          }
          print('  分类ID：$validCategoryIds');
        }
      }
      
      print('发送请求参数：$params');
      
      // 调用API
      final responseData = await _apiService.get(
        path: '/api/v1/finance/transaction-groups',
        params: params,
        context: context,
      );
      
      bool success = responseData['code'] == 0 || responseData['code'] == 200;
      
      return ApiResponse(
        success: success,
        data: success ? responseData['data'] : null,
        message: responseData['message'] ?? (success ? '获取成功' : '获取失败'),
      );
    } catch (e) {
      print("Error in FinanceService.getTransactionGroups: $e");
      return ApiResponse(
        success: false,
        message: '获取交易记录分组失败: $e',
      );
    }
  }
  
  /// 获取支出分析数据，按类别统计
  Future<ApiResponse> getExpenseAnalysis({
    required BuildContext context,
    DateTime? startDate,
    DateTime? endDate,
    int? memberId,
  }) async {
    try {
      // 构建查询参数
      Map<String, String> params = {};
      
      // 添加日期范围参数
      if (startDate != null) {
        params['start_date'] = DateFormat('yyyy-MM-dd HH:mm:ss').format(startDate);
      }
      if (endDate != null) {
        params['end_date'] = DateFormat('yyyy-MM-dd HH:mm:ss').format(endDate);
      }
      
      // 添加成员ID参数
      if (memberId != null) {
        params['member_id'] = memberId.toString();
      }
      
      print('获取支出分析数据，参数: $params');
      
      // 调用API
      final responseData = await _apiService.get(
        path: '/api/v1/finance/expense-analysis',
        params: params,
        context: context,
      );
      
      bool success = responseData['code'] == 0 || responseData['code'] == 200;
      
      return ApiResponse(
        success: success,
        data: success ? responseData['data'] : null,
        message: responseData['message'] ?? (success ? '获取成功' : '获取失败'),
      );
    } catch (e) {
      print("Error in FinanceService.getExpenseAnalysis: $e");
      return ApiResponse(
        success: false,
        message: '获取支出分析数据失败: $e',
      );
    }
  }

  // 获取家庭成员财务贡献数据
  Future<ApiResponse> getFamilyContributions({
    required BuildContext context,
    int? year,
    int? month,
  }) async {
    try {
      // 构建查询参数
      Map<String, String> params = {};
      if (year != null) params['year'] = year.toString();
      if (month != null) params['month'] = month.toString();
      
      // 发起请求
      final response = await _apiService.get(
        path: '/api/v1/finance/family-contributions',
        params: params,
        context: context,
      );
      
      // 返回响应
      bool success = response['code'] == 0 || response['code'] == 200;
      return ApiResponse(
        success: success,
        message: response['message'] ?? (success ? '获取成功' : '获取失败'),
        data: success ? response['data'] : null,
      );
    } catch (e) {
      debugPrint('获取家庭成员财务贡献数据失败: $e');
      return ApiResponse(
        success: false,
        message: '获取家庭成员财务贡献数据失败: $e',
        data: null,
      );
    }
  }
} 